panoptes_aggregation/panoptes/userify.py

```diff
@@ -0,0 +1,112 @@
+from collections import Iterable
+from json import dumps as jsonify
+from os import environ
+from flask import request
+from panoptes_client import Panoptes, User
+import requests
+
+
+class ConfigurationError(Exception):
+    pass
+
+
+known_params = [
+    'destination'
+]
+
+connected = False
+
+users = {}
+
+destinations = {
+    'mast': 'https://mast-forward.zooniverse.org/',
+    'mockable': 'https://demo1580318.mockable.io/mast'
+}
+
+
+def userify():
+    global users
+    if request.method == 'GET':
+        return 'Usage'
+    elif request.method in ['POST', 'PUT']:
+        find_fields = _discover_fields(request.args.to_dict())
+        target_object = request.get_json()
+        destination = request.args.get('destination')
+
+        _stuff_object(target_object, find_fields)
+        users = {}
+
+        if destination:
+            _forward_contents(target_object, destination)
+
+        return jsonify(target_object)
+
+
+def _forward_contents(payload, destination):
+    if destination not in destinations:
+        raise ConfigurationError('Unknown destination')
+
+    requests.post(url=destinations[destination], json=payload)
+    return
+
+
+def _stuff_object(target_object, find_fields):
+    for key in target_object.keys():
+        if type(target_object[key]) is dict:
+            _stuff_object(target_object[key], find_fields)
+
+    user_ids = _discover_user_ids(target_object)
+    for user_id in user_ids:
+        user = _retrieve_user(user_id)
+        target_object['users'] = target_object.get('users', [])
+        target_object['users'].append(_build_user_hash(user, find_fields))
+
+    return target_object
+
+
+def _discover_fields(request_args):
+    raw_list = request_args.keys()
+    return [i for i in raw_list if i not in known_params]
+
+
+def _discover_user_ids(target_object):
+    user_ids = []
+    if 'user_ids' in target_object:
+        user_ids = target_object['user_ids']
+    if 'user_id' in target_object:
+        user_ids.append(target_object['user_id'])
+
+    return _unique(_flatten(user_ids))
+
+
+def _build_user_hash(user, find_fields):
+    user_hash = {'id': user.id}
+    for key in find_fields:
+        user_hash[key] = getattr(user, key)
+    return user_hash
+
+
+def _unique(l):
+    return list(set(l))
+
+
+def _flatten(l):
+    for el in l:
+        if isinstance(el, Iterable) and not isinstance(el, (str, bytes)):
+            yield from _flatten(el)
+        else:
+            yield el
+
+
+def _retrieve_user(user_id):
+    global connected
+    if user_id in users:
+        user = users[user_id]
```

Comment:
Could this use a Redis cache store with autoexpiring keys after a time (1 day / week?) instead of a global var - possibly https://pythonhosted.org/Flask-Cache/#flask.ext.cache.Cache.memoize 

We can get you access to a Redis DB for this app. the user data won't really change so it would be nice to have an autoexpiring cache representation to cut down on the API traffic.

Reply:
the point of this app is to be a lightweight host for functions that keep very little state. i prefer to complicate it with as few dependencies as possible.

Reply:
YAGNI may definitely apply here. However how often will this app call the API and will it call it as a logged in admin user? 

I think each batch will generally miss your internal users cache but instead hit the API for every user_id in the list. If this is the case then this app will just be another vector of redundant traffic on the API. 

i'd prefer if we think about the the traffic patterns ahead of time for this function and cut down on slower API network traffic and instead save the costs / speed the app up by using a cache store where we can.

Reply:
if we want to build a caching server somewhere between this service and the API then that's probably a good idea but this app is, again, explicitly designed not to maintain state.